src/libcore/borrow.rs

```diff
@@ -12,26 +12,143 @@
+/// When writing generic code, a use of `Borrow` should always be justified
+/// by additional trait bounds, making it clear that the two types need to
+/// behave identically in a certain context. If the code should merely be
+/// able to operate on any type that can produce a reference to a given type,
+/// you should use [`AsRef`] instead.
+///
+/// The companion trait [`BorrowMut`] provides the same guarantees for
+/// mutable references.
+///
+/// [`Box<T>`]: ../../std/boxed/struct.Box.html
+/// [`Rc<T>`]: ../../std/rc/struct.Rc.html
+/// [`Vec<T>`]: ../../std/vec/struct.Vec.html
+/// [`AsRef`]: ../../std/convert/trait.AsRef.html
+/// [`BorrowMut`]: trait.BorrowMut.html
+///
+/// # Examples
+///
+/// As a data collection, [`HashMap<K, V>`] owns both keys and values. If
+/// the key’s actual data is wrapped in a managing type of some kind, it
+/// should, however, still be possible to search for a value using a
+/// reference to the key’s data. For instance, if the key is a string, then
+/// it is likely stored with the hash map as a [`String`], while it should
+/// be possible to search using a [`&str`][`str`]. Thus, `insert` needs to
+/// operate on a `String` while `get` needs to be able to use a `&str`.
+///
+/// Slightly simplified, the relevant parts of `HashMap<K, V>` look like
+/// this:
+///
+/// ```
+/// use std::borrow::Borrow;
+/// use std::hash::Hash;
+///
+/// pub struct HashMap<K, V> {
+///     # marker: ::std::marker::PhantomData<(K, V)>,
+///     // fields omitted
+/// }
+///
+/// impl<K, V> HashMap<K, V> {
+///     pub fn insert(&self, key: K, value: V) -> Option<V>
+///     where K: Hash + Eq
+///     {
+///         # unimplemented!()
+///         // ...
+///     }
+///
+///     pub fn get<Q>(&self, k: &Q) -> Option<&V>
+///     where
+///         K: Borrow<Q>,
+///         Q: Hash + Eq + ?Sized
+///     {
+///         # unimplemented!()
+///         // ...
+///     }
+/// }
+/// ```
+///
+/// The entire hash map is generic over a key type `K`. Because these keys
+/// are stored by with the hash map, this type as to own the key’s data.
+/// When inserting a key-value pair, the map is given such a `K` and needs
+/// to find the correct hash bucket and check if the key is already present
+/// based on that `K`. It therefore requires `K: Hash + Eq`.
+///
+/// In order to search for a value based on the key’s data, the `get` method
+/// is generic over some type `Q`. Technically, it needs to convert that `Q`
+/// into a `K` in order to use `K`’s [`Hash`] implementation to be able to
+/// arrive at the same hash value as during insertion in order to look into
+/// the right hash bucket. Since `K` is some kind of owned value, this likely
+/// would involve cloning and isn’t really practical.
+///
+/// Instead, `get` relies on `Q`’s implementation of `Hash` and uses `Borrow`
+/// to indicate that `K`’s implementation of `Hash` must produce the same
+/// result as `Q`’s by demanding that `K: Borrow<Q>`.
```

Comment:
I think there is a little more to it. If this were the end of the story, then a bound `Q: Borrow<K>` would work just as well. In fact that would seem to match the previous paragraph better -- `Q: Borrow<K>` means you can take the `&Q` and borrow a `&K` and use `K`'s Hash function to find the right bucket. Would it be possible to clarify what is going on and why `Q: Borrow<K>` is not just as good?

Reply:
I have rewritten the explanation, calling out what `Q` really is supposed to be. Is it more clear now?